# Praktikum Pengantar Pembelajaran Mesin


---
## Live Coding Bab 6. Support Vector Machine (SVM)

Nama: Hassan Nashrallah


NIM: 245150200111025


Kelas: F


Wine dataset adalah dataset klasik yang sering dipakai untuk **klasifikasi multi-kelas**. Dataset ini berisi **13 fitur numerik** tentang kandungan kimia dalam anggur (wine), seperti **alkohol, keasaman, magnesium, flavonoid, dll.**. Tujuannya adalah untuk mengklasifikasikan anggur menjadi **3 jenis berbeda** berdasarkan karakteristiknya.  

## Import library yang dibutuhkan

In [1]:
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split

## Load dan baca dataset

In [2]:
# Load Wine Dataset
wine = datasets.load_wine()
df = pd.DataFrame(wine.data, columns=wine.feature_names)
df['target'] = wine.target  # Add target column

Tampilkan 10 data pertama dan terakhir dari dataset

In [3]:
df.head(10)

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,target
0,14.23,1.71,2.43,15.6,127.0,2.80,3.06,0.28,2.29,5.64,1.04,3.92,1065.0,0
1,13.20,1.78,2.14,11.2,100.0,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050.0,0
2,13.16,2.36,2.67,18.6,101.0,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185.0,0
3,14.37,1.95,2.50,16.8,113.0,3.85,3.49,0.24,2.18,7.80,0.86,3.45,1480.0,0
4,13.24,2.59,2.87,21.0,118.0,2.80,2.69,0.39,1.82,4.32,1.04,2.93,735.0,0
5,14.20,1.76,2.45,15.2,112.0,3.27,3.39,0.34,1.97,6.75,1.05,2.85,1450.0,0
6,14.39,1.87,2.45,14.6,96.0,2.50,2.52,0.30,1.98,5.25,1.02,3.58,1290.0,0
7,14.06,2.15,2.61,17.6,121.0,2.60,2.51,0.31,1.25,5.05,1.06,3.58,1295.0,0
8,14.83,1.64,2.17,14.0,97.0,2.80,2.98,0.29,1.98,5.20,1.08,2.85,1045.0,0
9,13.86,1.35,2.27,16.0,98.0,2.98,3.15,0.22,1.85,7.22,1.01,3.55,1045.0,0


In [4]:
df.tail(10)

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,target
168,13.58,2.58,2.69,24.5,105.0,1.55,0.84,0.39,1.54,8.660000,0.74,1.80,750.0,2
169,13.40,4.60,2.86,25.0,112.0,1.98,0.96,0.27,1.11,8.500000,0.67,1.92,630.0,2
170,12.20,3.03,2.32,19.0,96.0,1.25,0.49,0.40,0.73,5.500000,0.66,1.83,510.0,2
171,12.77,2.39,2.28,19.5,86.0,1.39,0.51,0.48,0.64,9.899999,0.57,1.63,470.0,2
172,14.16,2.51,2.48,20.0,91.0,1.68,0.70,0.44,1.24,9.700000,0.62,1.71,660.0,2
173,13.71,5.65,2.45,20.5,95.0,1.68,0.61,0.52,1.06,7.700000,0.64,1.74,740.0,2
174,13.40,3.91,2.48,23.0,102.0,1.80,0.75,0.43,1.41,7.300000,0.70,1.56,750.0,2
175,13.27,4.28,2.26,20.0,120.0,1.59,0.69,0.43,1.35,10.200000,0.59,1.56,835.0,2
176,13.17,2.59,2.37,20.0,120.0,1.65,0.68,0.53,1.46,9.300000,0.60,1.62,840.0,2
177,14.13,4.10,2.74,24.5,96.0,2.05,0.76,0.56,1.35,9.200000,0.61,1.60,560.0,2


## Cek missing values

In [5]:
print(df.isnull().sum())

alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
target                          0
dtype: int64


## Normalisasi fitur
Dilarang menggunakan library

In [6]:
def normalisasi(df_input):
    df_input = df_input.copy()
    list_fitur = df_input.columns[ :- 1]
    for fitur in list_fitur:
        max = df_input[fitur].max()
        min = df_input[fitur].min()
        df_input[fitur]=(df_input[fitur]-min)/(max-min)
    return df_input

data_normal = normalisasi(df)
data_normal.head()

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,target
0,0.842105,0.191700,0.572193,0.257732,0.619565,0.627586,0.573840,0.283019,0.593060,0.372014,0.455285,0.970696,0.561341,0
1,0.571053,0.205534,0.417112,0.030928,0.326087,0.575862,0.510549,0.245283,0.274448,0.264505,0.463415,0.780220,0.550642,0
2,0.560526,0.320158,0.700535,0.412371,0.336957,0.627586,0.611814,0.320755,0.757098,0.375427,0.447154,0.695971,0.646933,0
3,0.878947,0.239130,0.609626,0.319588,0.467391,0.989655,0.664557,0.207547,0.558360,0.556314,0.308943,0.798535,0.857347,0
4,0.581579,0.365613,0.807487,0.536082,0.521739,0.627586,0.495781,0.490566,0.444795,0.259386,0.455285,0.608059,0.325963,0


## Bagi ke menjadi data uji dan data test
Diperbolehkan menggunakan library train_test_split

In [7]:
from sklearn.model_selection import train_test_split

data_latih, data_uji = train_test_split(data_normal, test_size=0.2, random_state=42)
# bagi untuk data latih menjadi 80% total dan dan data uji sisanya

label_latih = data_latih.pop('target')
label_uji = data_uji.pop('target')

## Fungsi SGD

In [8]:
from sklearn.utils import shuffle

def hitung_cost_gradient(w, x, y, regularization):
    jarak = 1 - (y * np.dot(x, w))
    if max(0, jarak) == 0:
        di = w
    else:
        di = w - (regularization * y * x)
    return di

def sgd(data_latih, label_latih, learning_rate=0.000001, max_epoch=1000, regularization=10000):
    data_latih = data_latih.to_numpy()
    label_latih = label_latih.to_numpy()
    bobot = np.zeros(data_latih.shape[1])
    for epoch in range(1,max_epoch):
        X, Y = shuffle(data_latih,label_latih, random_state=101)
        for index, x in enumerate(X):
            delta = hitung_cost_gradient(bobot, x, Y[index], regularization)
            bobot = bobot - (learning_rate * delta)
    return bobot

## Fungsi training

In [9]:
def buat_trainingset(data, label):
    dataset = data.copy(deep=True)
    dataset['target'] = label.values

    return {
        kelas: dataset.assign(target=dataset['target'].map({kelas: 1}).fillna(-1))
        for kelas in dataset['target'].unique()
    }

def training(trainingset):
    list_kelas = trainingset.keys()
    bobot = {}
    for kelas in list_kelas:
        data_temp = trainingset[kelas].copy() 
        label_latih = data_temp.pop(data_temp.columns[-1])
        bobot[kelas] = sgd(data_temp, label_latih)
    return bobot

## Fungsi testing

In [10]:
def testing(w, data_uji):
    kelas_prediksi = []
    for i in range(data_uji.shape[0]):
        x = data_uji.to_numpy()[i]
        hasil = {}
        for kelas, bobot in w.items():
            hasil[kelas] = np.sign(np.dot(bobot, x))
        kandidat = [kelas for kelas, nilai in hasil.items() if nilai == 1]
        if len(kandidat) == 1:
            kelas_prediksi.append(kandidat[0])
        else:
            skor = {kelas: np.dot(bobot, x) for kelas, bobot in w.items()}
            kelas_prediksi.append(max(skor, key=skor.get))
    return kelas_prediksi

## Gunakan fungsi yang sudah dibuat

In [11]:
trainingset = buat_trainingset(data_latih, label_latih)
W = training(trainingset)
prediksi = testing(W, data_uji)

## Hitung akurasi

In [12]:
print(f'Accuracy: {np.mean(prediksi == label_uji)*100:.2f}%') 
# isi bagian {...} untuk menghitung akurasi

Accuracy: 97.22%


## Penjelasan dan kesimpulan

Live coding ini mengimplementasikan algoritma Support Vector Machine (SVM) dari nol untuk menyelesaikan masalah klasifikasi multi-class pada dataset wine. Proses dimulai dari eksplorasi dan pengecekan data, dilanjutkan dengan normalisasi fitur menggunakan min-max scaling agar semua fitur berada pada skala yang sama. Data kemudian dibagi menjadi data latih dan data uji, dan model SVM dilatih menggunakan metode SGD. Untuk menangani multi-class, digunakan pendekatan One-vs-Rest, di mana setiap kelas dilatih sebagai model biner terpisah. Setelah itu, model diuji dengan memilih kelas berdasarkan skor tertinggi dari masing-masing model, dan performanya dievaluasi menggunakan skor akurasi.

Dari skor akurasi yang didapatkan, model SVM ini memiliki performa yang sangat baik dengan skor akurasi mencapai 97.22%. Hal ini berarti model SVM berhasil memprediksi hampir seluruh data uji dengan benar.
